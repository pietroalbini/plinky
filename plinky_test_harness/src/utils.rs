use anyhow::{bail, Error};
use std::path::{Path, PathBuf};
use std::process::Command;

pub fn run(command: &mut Command) -> Result<(), Error> {
    let cmd_repr = format!("{command:?}");
    let output = command.output()?;
    if !output.status.success() {
        eprintln!("Failed to execute {cmd_repr}");
        eprintln!();
        eprintln!("=== stdout ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stdout));
        eprintln!();
        eprintln!("=== stderr ===");
        eprintln!("{}", String::from_utf8_lossy(&output.stderr));
        eprintln!();
        bail!("command failed with exit {}", output.status);
    }
    Ok(())
}

pub struct RunAndSnapshot {
    name: String,
    path: PathBuf,
    output: String,
}

impl RunAndSnapshot {
    pub fn new(name: &str, path: &Path) -> Self {
        Self { name: name.into(), path: path.into(), output: String::new() }
    }

    pub fn run(&mut self, action: &str, command: &mut Command) -> Result<bool, Error> {
        self.separator();
        match command.output() {
            Ok(output) => {
                self.output.push_str(&format!("{action} exited with {}\n", output.status));
                for (name, content) in [("stdout", &output.stdout), ("stderr", &output.stderr)] {
                    if content.trim_ascii().is_empty() {
                        self.output.push_str(&format!("\nno {name} present\n"));
                    } else {
                        let mut content = String::from_utf8_lossy(content).to_string();
                        content = content.replace(env!("CARGO_MANIFEST_DIR"), "${project}");

                        self.output
                            .push_str(&format!("\n=== {name} ===\n{}\n", content.trim_ascii_end()));
                    }
                }
                Ok(output.status.success())
            }
            Err(err) => {
                self.output.push_str(&format!("{action} failed to execute with error: {err}"));
                Ok(false)
            }
        }
    }

    pub fn note(&mut self, note: &str) {
        self.separator();
        self.output.push_str(note);
        if !note.ends_with('\n') {
            self.output.push('\n');
        }
    }

    fn separator(&mut self) {
        if !self.output.is_empty() {
            self.output.push_str("\n==============\n\n");
        }
    }

    pub fn persist(self) {
        let mut insta_settings = insta::Settings::clone_current();
        insta_settings.set_prepend_module_to_snapshot(false);
        insta_settings.set_omit_expression(true);
        insta_settings.set_snapshot_path(self.path.canonicalize().unwrap());

        insta_settings.bind(|| {
            insta::assert_snapshot!(self.name, self.output.trim_ascii_end());
        });
    }
}

pub(crate) fn file_name(path: impl AsRef<Path>) -> String {
    path.as_ref().file_name().unwrap().to_str().unwrap().to_string()
}

#[track_caller]
pub(crate) fn err_str<T>(result: Result<T, Error>) -> Result<T, String> {
    match result {
        Ok(ok) => Ok(ok),
        Err(err) => {
            let mut repr = format!("error: {err}\n");
            let mut source = err.source();
            while let Some(err) = source {
                repr.push_str(&format!("  cause: {err}\n"));
                source = err.source();
            }
            panic!("{repr}")
        }
    }
}
