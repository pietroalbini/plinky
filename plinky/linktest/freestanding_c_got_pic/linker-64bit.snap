---
source: plinky_test_harness/src/utils.rs
---
linking exited with exit status: 0

no stdout present

=== stderr ===
debug print: loaded object
 │
 │  section .text#1 (perms: rx) in hello.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 48 89 e5 ba 0c 00 00 00 48 8d 05 00 00 00 00 │ UH.......H...... │
 │   │  │ 48 89 c6 bf 01 00 00 00 ff 15 00 00 00 00 bf 00 │ H............... │
 │   │  │ 00 00 00 ff 15 00 00 00 00                      │ .........        │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   │
 │   │    Relocations:
 │   │  ╭───────────────┬─────────────────────┬────────┬────────────────────╮
 │   │  │ Type          │ Symbol              │ Offset │ Addend             │
 │   │  ├───────────────┼─────────────────────┼────────┼────────────────────┤
 │   │  │ Relative32    │ <section .rodata#5> │ 0xc    │ 0xfffffffffffffffc │
 │   │  ├───────────────┼─────────────────────┼────────┼────────────────────┤
 │   │  │ GOTRelative32 │ write#8             │ 0x1a   │ 0xfffffffffffffffc │
 │   │  ├───────────────┼─────────────────────┼────────┼────────────────────┤
 │   │  │ GOTRelative32 │ exit#9              │ 0x25   │ 0xfffffffffffffffc │
 │   │  ╰───────────────┴─────────────────────┴────────┴────────────────────╯
 │   ┴
 │
 │  section .text#15 (perms: rx) in syscalls.o
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 53 51 48 c7 c0 04 00 00 00 48 89 fb 48 89 f1 │ PSQH......H..H.. │
 │   │  │ cd 80 59 5b 58 c3 48 c7 c0 01 00 00 00 48 89 fb │ ..Y[X.H......H.. │
 │   │  │ cd 80                                           │ ..               │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 ┴

debug print: calculated layout
 │
 │    Sections:
 │  ╭─────────────┬─────────────────────────────┬────────────────╮
 │  │ Section     │ Source object               │ Memory address │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .interp#21  │ <plinky>                    │ 0x1000         │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .rodata#5   │ hello.o                     │ 0x2000         │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .eh_frame#8 │ hello.o                     │ 0x200d         │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .text#1     │ hello.o                     │ 0x3000         │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .text#15    │ syscalls.o                  │ 0x3029         │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .got#23     │ <plinky>                    │ 0x4000         │
 │  ├─────────────┼─────────────────────────────┼────────────────┤
 │  │ .comment#22 │ mix of hello.o and <plinky> │ not allocated  │
 │  ╰─────────────┴─────────────────────────────┴────────────────╯
 │
 │    Segments:
 │  ╭────────┬────────┬─────────────┬────────────────────┬─────────────╮
 │  │ Start  │ Align  │ Type        │ Permissions        │ Content     │
 │  ├────────┼────────┼─────────────┼────────────────────┼─────────────┤
 │  │ 0x1000 │ 0x1000 │ interpreter │ ElfPermissions(R)  │ .interp#21  │
 │  ├────────┼────────┼─────────────┼────────────────────┼─────────────┤
 │  │ 0x2000 │ 0x1000 │ program     │ ElfPermissions(R)  │ .rodata#5   │
 │  │        │        │             │                    │ .eh_frame#8 │
 │  ├────────┼────────┼─────────────┼────────────────────┼─────────────┤
 │  │ 0x3000 │ 0x1000 │ program     │ ElfPermissions(RX) │ .text#1     │
 │  │        │        │             │                    │ .text#15    │
 │  ├────────┼────────┼─────────────┼────────────────────┼─────────────┤
 │  │ 0x4000 │ 0x1000 │ program     │ ElfPermissions(RW) │ .got#23     │
 │  ╰────────┴────────┴─────────────┴────────────────────┴─────────────╯
 │
 │    deduplication facade .comment#6 in hello.o
 │  ╭──────┬───────────────────╮
 │  │ From │ To                │
 │  ├──────┼───────────────────┤
 │  │ 0x0  │ .comment#22 + 0x0 │
 │  ├──────┼───────────────────┤
 │  │ 0x1  │ .comment#22 + 0x1 │
 │  ╰──────┴───────────────────╯
 │
 │    deduplication facade .comment#13 in <plinky>
 │  ╭──────┬────────────────────╮
 │  │ From │ To                 │
 │  ├──────┼────────────────────┤
 │  │ 0x0  │ .comment#22 + 0x26 │
 │  ╰──────┴────────────────────╯
 ┴

debug print: object after relocations are applied
 │
 │  section .got#23 (perms: rw) in <plinky>
 │   │
 │   │  address: 0x4000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 │ ................ │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .interp#21 (perms: r) in <plinky>
 │   │
 │   │  address: 0x1000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 2f 6c 69 62 36 34 2f 6c 64 2d 6c 69 6e 75 78 2d │ /lib64/ld-linux- │
 │   │  │ 78 38 36 2d 36 34 2e 73 6f 2e 32 00             │ x86-64.so.2.     │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#1 (perms: rx) in hello.o
 │   │
 │   │  address: 0x3000
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 55 48 89 e5 ba 0c 00 00 00 48 8d 05 f0 ef ff ff │ UH.......H...... │
 │   │  │ 48 89 c6 bf 01 00 00 00 ff 15 e2 0f 00 00 bf 00 │ H............... │
 │   │  │ 00 00 00 ff 15 df 0f 00 00                      │ .........        │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .text#15 (perms: rx) in syscalls.o
 │   │
 │   │  address: 0x3029
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 50 53 51 48 c7 c0 04 00 00 00 48 89 fb 48 89 f1 │ PSQH......H..H.. │
 │   │  │ cd 80 59 5b 58 c3 48 c7 c0 01 00 00 00 48 89 fb │ ..Y[X.H......H.. │
 │   │  │ cd 80                                           │ ..               │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │    Symbols:
 │  ╭─────────────────────────┬──────────┬────────────┬────────────┬─────────────────╮
 │  │ Name                    │ Type     │ Source     │ Visibility │ Value           │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ <symbol#0>              │ none     │ <plinky>   │ local      │ <null>          │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ <section .text#1>       │ section  │ hello.o    │ local      │ .text#1 + 0x0   │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ <section .rodata#5>     │ section  │ hello.o    │ local      │ .rodata#5 + 0x0 │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ _GLOBAL_OFFSET_TABLE_#7 │ none     │ hello.o    │ global     │ <undefined>     │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ _start#1                │ function │ hello.o    │ global     │ .text#1 + 0x0   │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ exit#9                  │ none     │ syscalls.o │ global     │ .text#15 + 0x16 │
 │  ├─────────────────────────┼──────────┼────────────┼────────────┼─────────────────┤
 │  │ write#8                 │ none     │ syscalls.o │ global     │ .text#15 + 0x0  │
 │  ╰─────────────────────────┴──────────┴────────────┴────────────┴─────────────────╯
 │
 │    Dynamic symbols:
 │  ╭────────────┬──────┬────────────┬────────────┬─────────────────╮
 │  │ Name       │ Type │ Source     │ Visibility │ Value           │
 │  ├────────────┼──────┼────────────┼────────────┼─────────────────┤
 │  │ <symbol#0> │ none │ <plinky>   │ local      │ <null>          │
 │  ├────────────┼──────┼────────────┼────────────┼─────────────────┤
 │  │ exit#9     │ none │ syscalls.o │ global     │ .text#15 + 0x16 │
 │  ├────────────┼──────┼────────────┼────────────┼─────────────────┤
 │  │ write#8    │ none │ syscalls.o │ global     │ .text#15 + 0x0  │
 │  ╰────────────┴──────┴────────────┴────────────┴─────────────────╯
 │
 │    Dynamic relocations:
 │  ╭─────────────┬─────────┬────────┬────────╮
 │  │ Type        │ Symbol  │ Offset │ Addend │
 │  ├─────────────┼─────────┼────────┼────────┤
 │  │ FillGOTSlot │ write#8 │ 0x4000 │ 0x0    │
 │  ├─────────────┼─────────┼────────┼────────┤
 │  │ FillGOTSlot │ exit#9  │ 0x4008 │ 0x0    │
 │  ╰─────────────┴─────────┴────────┴────────╯
 ┴

debug print: built elf
 │
 │    Metadata:
 │  ╭────────────┬───────────────╮
 │  │ Class      │ ELF 64bit     │
 │  ├────────────┼───────────────┤
 │  │ Endian     │ Little        │
 │  ├────────────┼───────────────┤
 │  │ ABI        │ System V      │
 │  ├────────────┼───────────────┤
 │  │ Machine    │ x86-64        │
 │  ├────────────┼───────────────┤
 │  │ Type       │ Shared object │
 │  ├────────────┼───────────────┤
 │  │ Entrypoint │ 0x3000        │
 │  ╰────────────┴───────────────╯
 │
 │  section .interp#6 (address: 0x1000)
 │   │
 │   │  program data | permissions: R
 │   │
 │   │  ╭─────────────────────────────────────────────────┬──────────────────╮
 │   │  │ 2f 6c 69 62 36 34 2f 6c 64 2d 6c 69 6e 75 78 2d │ /lib64/ld-linux- │
 │   │  │ 78 38 36 2d 36 34 2e 73 6f 2e 32 00             │ x86-64.so.2.     │
 │   │  ╰─────────────────────────────────────────────────┴──────────────────╯
 │   ┴
 │
 │  section .dynstr#9 (address: 0x5000)
 │   │
 │   │    Strings table:
 │   │  ╭─────┬───────╮
 │   │  │ 0x0 │       │
 │   │  ├─────┼───────┤
 │   │  │ 0x1 │       │
 │   │  ├─────┼───────┤
 │   │  │ 0x2 │ write │
 │   │  ├─────┼───────┤
 │   │  │ 0x8 │ exit  │
 │   │  ╰─────┴───────╯
 │   ┴
 │
 │  section .dynsym#10 (address: 0x500d)
 │   │
 │   │    Dynamic symbol table:
 │   │  ╭─────────┬─────────┬──────┬────────────┬────────────┬────────┬──────╮
 │   │  │ Name    │ Binding │ Type │ Visibility │ Definition │ Value  │ Size │
 │   │  ├─────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ #0      │ Local   │ -    │ Default    │ Undefined  │ 0x0    │ 0x0  │
 │   │  ├─────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ write#1 │ Global  │ -    │ Default    │ .text#5    │ 0x3029 │ 0x0  │
 │   │  ├─────────┼─────────┼──────┼────────────┼────────────┼────────┼──────┤
 │   │  │ exit#2  │ Global  │ -    │ Default    │ .text#5    │ 0x303f │ 0x0  │
 │   │  ╰─────────┴─────────┴──────┴────────────┴────────────┴────────┴──────╯
 │   ┴
 │
 │  section .rela.dyn#11 (address: 0x5055)
 │   │
 │   │  symbol table:       .dynsym#10
 │   │  applies to section: #0
 │   │
 │   │    Relocations:
 │   │  ╭────────────────┬─────────┬────────┬────────╮
 │   │  │ Type           │ Symbol  │ Offset │ Addend │
 │   │  ├────────────────┼─────────┼────────┼────────┤
 │   │  │ X86_64_GlobDat │ write#1 │ 0x4000 │ 0x0    │
 │   │  ├────────────────┼─────────┼────────┼────────┤
 │   │  │ X86_64_GlobDat │ exit#2  │ 0x4008 │ 0x0    │
 │   │  ╰────────────────┴─────────┴────────┴────────╯
 │   ┴
 │
 │  section .hash#12 (address: 0x5085)
 │   │
 │   │  Hash table for .dynsym#10
 │   │
 │   │    Content:
 │   │  ╭───────────┬───────────────────╮
 │   │  │ Bucket ID │ Symbols in bucket │
 │   │  ├───────────┼───────────────────┤
 │   │  │ 0         │ exit#2            │
 │   │  │           │ write#1           │
 │   │  ╰───────────┴───────────────────╯
 │   ┴
 │
 │  section .dynamic#13 (address: 0x509d)
 │   │
 │   │  dynamic information | string table: .dynstr#9
 │   │
 │   │  ╭─────────────────────────┬────────────────╮
 │   │  │ Kind                    │ Value          │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Hash table              │ address 0x5085 │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ String table            │ address 0x5000 │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ String table size       │ 13 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Symbol table            │ address 0x500d │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Symbol table entry size │ 24 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ RelocationsA table      │ address 0x5055 │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ RelocationsA size       │ 48 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ RelocationsA entry size │ 24 bytes       │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Flags1                  │ pie            │
 │   │  ├─────────────────────────┼────────────────┤
 │   │  │ Null                    │ -              │
 │   │  ╰─────────────────────────┴────────────────╯
 │   ┴
 │
 │    Segments:
 │  ╭──────────────────────┬───────┬──────────┬────────────────────────────────────────────────────────────╮
 │  │ Type                 │ Perms │ Aligment │ Content                                                    │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Program header table │ R     │ 0x1000   │ program header                                             │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Interpreter          │ R     │ 0x1000   │ .interp#6                                                  │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Load                 │ R     │ 0x1000   │ program header                                             │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Load                 │ R     │ 0x1000   │ elf header                                                 │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Load                 │ R     │ 0x1000   │ .rodata#3, .eh_frame#4                                     │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Load                 │ R X   │ 0x1000   │ .text#2, .text#5                                           │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Load                 │ RW    │ 0x1000   │ .got#8                                                     │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Load                 │ R     │ 0x1000   │ .dynstr#9, .dynsym#10, .rela.dyn#11, .hash#12, .dynamic#13 │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ Dynamic              │ R     │ 0x8      │ .dynamic#13                                                │
 │  ├──────────────────────┼───────┼──────────┼────────────────────────────────────────────────────────────┤
 │  │ GNU stack            │ RW    │ 0x1      │ -                                                          │
 │  ╰──────────────────────┴───────┴──────────┴────────────────────────────────────────────────────────────╯
 ┴
